pkg/volume/csi/csi_attacher.go

```diff
@@ -488,30 +465,31 @@ func (c *csiAttacher) waitForVolumeDetachmentInternal(volumeHandle, attachID str
 		case event, ok := <-ch:
 			if !ok {
 				klog.Errorf("[attachment.ID=%v] watch channel had been closed", attachID)
-				return errors.New("volume attachment watch channel had been closed")
+				return "", errors.New("volume attachment watch channel had been closed")
 			}
 
 			switch event.Type {
 			case watch.Added, watch.Modified:
 				attach, _ := event.Object.(*storage.VolumeAttachment)
-				// driver reports attach error
-				detachErr := attach.Status.DetachError
-				if detachErr != nil {
-					klog.Error(log("detachment for VolumeAttachment [%v] for volume [%s] failed: %v", attachID, volumeHandle, detachErr.Message))
-					return errors.New(detachErr.Message)
+				successful, err := waitForStatus(attach, volumeHandle)
+				if err != nil {
+					return "", err
+				}
+				if successful {
+					return attachID, nil
 				}
 			case watch.Deleted:
 				//object deleted
 				klog.V(4).Info(log("VolumeAttachment object [%v] for volume [%v] has been deleted", attachID, volumeHandle))
-				return nil
+				return attachID, nil
```

Comment:
This is correct when we wait for detach (i.e. VolumeAttachment being deleted). It should be error when we're waiting for attach.

Reply:
There is still a bug, IMO. Current code in `case watch.Deleted:` works for detachment - a volume has been detached when VolumeAttachment is deleted.

It does not work for attachment - something went wrong when `VolumeAttachment` got deleted and the volume is definitely not attached. It must not return success in this case.

* You can either add extra parameter to `waitForVolumeAttachmentInternal` to differentiate between attach/detach and return proper return value on delete event .
* Or you can run `verifyStatus(nil, volumeHandle)`, where `nil` means that the attachment has been deleted (don't forget to document this).
* Or you can find a better solution, of course.

Reply:
I chose the 2rd method and added some comments.

PTAL, thanks